L1Trigger/TrackFindingAM/src/RetinaTrackFitter.cc

```diff
@@ -0,0 +1,754 @@
+
+  // --- Conformal transformation: 
+  confTrans(hits);
+
+  //
+  // --- First step ------------------------------------------------------------
+  //
+
+  // --- Setup the retina:
+  double pbins_step1 = config[trigTow_type]["xy_pbins_step1"];
+  double qbins_step1 = config[trigTow_type]["xy_qbins_step1"];
+  double pmin_step1  = config[trigTow_type]["xy_pmin_step1"];
+  double pmax_step1  = config[trigTow_type]["xy_pmax_step1"];
+  double qmin_step1  = config[trigTow_type]["xy_qmin_step1"];
+  double qmax_step1  = config[trigTow_type]["xy_qmax_step1"];
+  
+  double minWeight_step1 = config[trigTow_type]["xy_threshold_step1"];
+
+  double pstep_step1 = (pmax_step1-pmin_step1)/pbins_step1;
+  double qstep_step1 = (qmax_step1-qmin_step1)/qbins_step1;
+
+  vector <double> sigma_step1(8,sqrt(pstep_step1*pstep_step1+qstep_step1*qstep_step1));
+  if ( config[trigTow_type]["xy_sigma1_step1"] != 0. ) 
+    sigma_step1[0] = config[trigTow_type]["xy_sigma1_step1"];
+  if ( config[trigTow_type]["xy_sigma2_step1"] != 0. ) 
+    sigma_step1[1] = config[trigTow_type]["xy_sigma2_step1"];
+  if ( config[trigTow_type]["xy_sigma3_step1"] != 0. ) 
+    sigma_step1[2] = config[trigTow_type]["xy_sigma3_step1"];
+  if ( config[trigTow_type]["xy_sigma4_step1"] != 0. ) 
+    sigma_step1[3] = config[trigTow_type]["xy_sigma4_step1"];
+  if ( config[trigTow_type]["xy_sigma5_step1"] != 0. ) 
+    sigma_step1[4] = config[trigTow_type]["xy_sigma5_step1"];
+  if ( config[trigTow_type]["xy_sigma6_step1"] != 0. ) 
+    sigma_step1[5] = config[trigTow_type]["xy_sigma6_step1"];
+  if ( config[trigTow_type]["xy_sigma7_step1"] != 0. ) 
+    sigma_step1[6] = config[trigTow_type]["xy_sigma7_step1"];
+  if ( config[trigTow_type]["xy_sigma8_step1"] != 0. ) 
+    sigma_step1[7] = config[trigTow_type]["xy_sigma8_step1"];
+
+
+  Retina retinaXY_step1(hits, pbins_step1+2, qbins_step1+2, 
+			pmin_step1-pstep_step1, pmax_step1+pstep_step1, 
+			qmin_step1-qstep_step1, qmax_step1+qstep_step1, 
+			sigma_step1, minWeight_step1, 1, XY);
+
+
+  // --- Fill the retina and find maxima:
+  retinaXY_step1.fillGrid();
+  retinaXY_step1.findMaxima();
+  if ( verboseLevel==1 )
+    retinaXY_step1.dumpGrid(event_counter,1,0);
+  if ( verboseLevel==2 )
+    retinaXY_step1.printMaxima();
+
+
+  // --- Get first step maxima:
+  vector <pqPoint> maximaXY_step1 = retinaXY_step1.getMaxima();
+
+  if ( maximaXY_step1.size()==0 && verboseLevel>0 ){
+    cout << "*** WARNING in RetinaTrackFitter::fit() at event/road = " << event_counter << "/" 
+	 << road_id << ": no maximum found in XY fit-step 1." << endl;
+    retinaXY_step1.dumpGrid(event_counter,1,0);
+  }
+
+  if ( maximaXY_step1.size() > 10 ){
+    cout << "*** ERROR in RetinaTrackFitter::fit() at event/road = " << event_counter 
+	 << "/" << road_id << ": " << maximaXY_step1.size() 
+	 << " maxima found in XY fit-step 1, fit aborted!" << endl;
+    if ( verboseLevel>0 ) 
+      retinaXY_step1.dumpGrid(event_counter,1,0);
+    return;
```

Comment:
If the function returns here the members in `hits` (`new`ed on line 70) are never freed.